Step2 - 지하철 노선 기능

aksgudwns opened the pull request:
안녕하세요 
지하철 노선 기능 리뷰 부탁드립니다.

지난번에 말씀주신 리팩토링 내용도 이번에 반영해보았습니다:)
다음부턴 요구사항 놓치지 않도록 꼼꼼히 챙기겠습니다!

감사합니다!

aksgudwns committed:
refactor - step1리뷰 반영

- 지하철역 조회하는 로직 공통 메소드로 분리

aksgudwns committed:
refactor - 각각의 인수테스트 격리

aksgudwns committed:
doc - 요구사항 추가

aksgudwns committed:
feat - 지하철 생성 ATDD, 기능 구현

aksgudwns committed:
fix - JsonIgnoreProperties Serialize안되는 오류 수정

aksgudwns committed:
feat - 지하철 목록 조회 기능 구현

- 목록조회 기능 구현
- 지하철 두개 생성 후 목록 조회하는 atdd 작성
- 생성 후 목록조회에 생성한 지하철역이 조회되는지 체크하는 로직추가
- 엔티티를 ResponseDto와 분리

aksgudwns committed:
feat - 지하철 조회 기능 구현

- 지하철 조회 기능 구현
- 지하철 조회 atdd작성
- 테스트메소드 이름 변경 retrieveLines -> retrieveAllLines
- 지하철호선, 지하철역 관계 재설정
- controller, service 잘못된 이름 수정

aksgudwns committed:
doc - 지하철노선 조회 구현한 기능 체크

aksgudwns committed:
feat - 지하철정보 수정 기능 구현

aksgudwns committed:
feat - 지하철 삭제 기능 구현

seokhongkim commented on README.md:
```diff
@@ -52,3 +52,32 @@ npm run dev
+ - 인수 테스트를 충족하는 기능 구현
+* 인수 테스트의 결과가 다른 인수테스트에 영향을 끼치지 않도록 인수 테스트를 서로 격리 시키세요.
+* 인수 테스트의 재사용성과 가독성, 그리고 빠른 테스트의도 파악을 위해 인수 테스트를 리팩터링 하세요.
+* [] 지하철노선 생성 
```
마크다운에서 체크리스트를 표현하는 문법인 `- [ ]` 로 사용하시고, 완료된 작업은 `- [x]` 로 표시하시면 문서 작성하실 때 도움이 될 것 같습니다. :)

aksgudwns replied:
완료작업이 x였군요.. 수정하겠습니다!

seokhongkim commented on src/test/java/nextstep/subway/DatabaseCleaner.java:
```diff
@@ -0,0 +1,41 @@
+    public void execute() {
+        entityManager.createNativeQuery("SET REFERENTIAL_INTEGRITY FALSE").executeUpdate();
+        tableNames.stream().forEach(tableName -> {
+            entityManager.createNativeQuery("TRUNCATE TABLE " + tableName).executeUpdate();
```
테이블 초기화를 통한 테스트 격리 구현 좋습니다. 👍

seokhongkim commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,242 @@
+        Long downStationId = Long.parseLong(StationAcceptanceTest.createStationAndGetId("새로운지하철역"));
+
+        //when
+        assertThat(createLineAndGetResponseCode(
```
when 구간에서는 API 호출을 하고, 검증은 then 구간에서 해보시면 좋을 것 같습니다. :)
given - when - then 에서는 when 은 "특정 행위를 할 때" 를 나타내므로, 검증은 "~이어야 한다"에 해당하는 then 구간에서 하시는 것을 추천합니다.

aksgudwns replied:
넵 수정하겠습니다!

seokhongkim commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,242 @@
+        //then
+        assertThat(lineNames.size()).isEqualTo(2);
+        assertThat(lineNames.stream()
+                .allMatch(lineName -> lineName1.equals(lineName) || lineName2.equals(lineName)))
```
정확한 검증을 위해서는 `lineNames` 가 정확히 "신분당선", "분당선" 을 가지고 있는지 검증하는 것이 좋겠습니다.
`createLine` 에서 전달하는 lineName 이 모두 `lineName1` 인 경우에도 생성만 된다면, 이 테스트는 통과할 수 있을 것 같습니다.

aksgudwns replied:
filter와 Set을 활용하여 수정해보겠습니다!

seokhongkim commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,242 @@
+                .isEqualTo(HttpStatus.CREATED.value());
+
+        //then
+        assertThat(retrieveAllLineNames().stream().anyMatch(lineName::equals))
```
생성이 완료되면 오는 응답 데이터를 이용해서 검증을 해보는 건 어떨까요?
생성도 API로 요청하고, 조회도 API로 요청하니 문제가 생겼을 때 두 API 중 무엇이 문제인지 다시 확인이 필요할 것 같아요!

aksgudwns replied:
넵 수정하겠습니다!

seokhongkim commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,242 @@
+                new LineRequest(lineName, "bg-red-600", upStationId, downStationId, 10));
+
+        //when
+        deleteLine(lineId);
```
여기도 삭제 이후 응답으로 오는 내용을 검증해보셔도 좋겠습니다. :)

aksgudwns replied:
같이 수정하였습니다!

seokhongkim commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -23,139 +26,98 @@ public class StationAcceptanceTest {
      */
-    @DisplayName("지하철역을 생성한다.")
     @Test
-    void createStation() {
```
리팩토링 좋습니다. 👍

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.stream.Collectors;
+
+@Service
+@Transactional(readOnly = true)
```
클래스에 설정한 트랜잭션의 readOnly 속성을 true 로 설정하신 이유가 있을까요?

aksgudwns replied:
readOnly true로 할 경우 클래스안에 @Transactional이 붙지 않은 메서드들은 조회용, @Transactional이 붙은 메소드들은 CUD작업이 일어나는 것을 명시적으로 보여 준다고 생각하였습니다..
또 readOnly true일 경우 영속성 컨텍스트가 스냅샷을 저장하지 않아서 성능상의 이점이 있는 것으로 알고 있습니다

하지만 찾아보니 DB vendor사나 버전에 따라 readOnly속성이 동작 안하는 경우도 있다 하고, 
JpaRepository에서 제공하는 메서드들은 Transactional true, false가 목적에 맞게 설정되어 있는 것으로 확인하였습니다.. 
메서드명으로 c,r,ud는 명시적으로 보여주고 있고, 성능상의 이점도 jpaRepository 메서드에 이미 어노테이션 설정되어 있으니 의미가 없어보이네요.. 

지금은 서비스 메서드 안에 cud가 일어는 Jpa repository api 호출이 하나 밖에 없어서 @Transactional이 의미 없어 보이지만, 
나중에 하나의 메서드 안에 cud가 일어나는 jpa repository api 호출을 여러번 할 경우 트랜잭션 관리가 필요할 수 있으므로,
class에 @Transactional만 붙이도록 수정하겠습니다.

혹시 제가 잘못이해한 부분있다면 피드백 부탁드립니다!

seokhongkim replied:
1. jpaRepository 메서드에 `@Transactional`이 설정되어 있지만, 여기에서 사용하는 전파레벨이 기본값인 `Propagation.REQUIRED`이므로 서비스에서 설정한 `readOnly` 설정을 그대로 따르게 되어 있습니다. 따라서 jpaRepository에 설정된 트랜잭션과 별개로 실제로 사용을 원하는 트랜잭션 속성을 서비스에서 지정해주실 필요는 있습니다.

2. 잘 아시겠지만 코드 블록을 \`(backtick) 문자로 감싸서 코드로 표현해주셔야 어노테이션을 나타내는 `@`가 멘션으로 동작하지 않습니다. `transactional` 계정이 멘션된 것으로 보여 말씀드립니다. :)

3. 저는 클래스의 transactional 설정은 readOnly = false 를 사용하는데요. 이유는 아래와 같습니다.
아래 내용도 함께 고민해보시면 도움이 될 것 같습니다. :)
* transactional 설정이 되었음을 테스트에서 확인하기 어려움. 즉, 개발자 또는 리뷰어가 놓치는 상황이 발생할 수 있음.
* readOnly 설정을 DB replication 과 함께 사용하여 writer/reader 를 구분하는 목적으로 사용할 수 있는데, readOnly = true 가 기본인 경우 쓰기 요청이 읽기 DB로 전달되어 장애 상황이 발생할 수 있음. 반면, 읽기 전용으로 설정해야 하는 작업이 쓰기 DB로 전달되는 경우 부하는 늘어날 수 있으나 당장 장애 상황으로 이어지지는 않음.
* 코드에서의 검증은 단위테스트와 통합테스트를 통해 검증이 상대적으로 용이함.

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,63 @@
+
+    private Line lineRequestToLine(LineRequest lineRequest) {
+        return new Line(lineRequest.getName(), lineRequest.getColor(),
+                stationRepository.getById(lineRequest.getUpStationId()),
```
1. `stationRepository.getById()` 로 존재하지 않는 ID로 데이터를 요청할 때 무슨 응답이 오는지 확인해보시면 좋겠습니다.
2. 존재하지 않는 역으로 노선을 만드는 경우에 적절한 예외 처리가 필요할 것 같습니다.
3. `stationRepository.getById()` 로 데이터를 조회하는 부분을 별도의 변수로 추출해보시면 좋겠습니다. 또는 메서드 밖에서 `Station` 조회 이후 전달해주면 `lineRequestToLine` 메서드의 역할을 하나(타입 변경)로 유지할 수도 있을 것 같습니다.

aksgudwns replied:
getById로 존재하지 않는 ID 조회시 DB Exception이 발생합니다..
적절한 예외처리를 위해 findById로 조회 후 데이터가 없는 경우 Exception 발생시키겠습니다.
또 메서드 역할을 하나로 유지하기 위해 분리작업도 진행해보았습니다!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Transactional
+    public void updateLine(Long id, LineUpdateRequest lineUpdateRequest) {
```
API 스펙에 따르면 지하철 노선 수정은 노선 이름과 색상만 변경하는 것으로 보이네요!
> https://edu.nextstep.camp/s/urD8knTu/ls/rnruA9Nq
`API 명세` 하단의 접기/펼치기 섹션을 눌러서 참고 부탁드립니다.

aksgudwns replied:
LineUpdateRequest는 name과 color만 있는 객체입니다!
다른방법으로 수정해야 할까요?

seokhongkim replied:
앗 제가 코드를 잘 못 읽었네요. 😓 
새로운 라인을 만들 때 `line.getUpStation()`을 가져오는 것을 `lineUpdateRequest`에서 꺼내서 업데이트 하는 것으로 이해했습니다.

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,63 @@
+    public void updateLine(Long id, LineUpdateRequest lineUpdateRequest) {
+        Line line = lineRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException("주어진 id로 생성된 지하철호선이 없습니다."));
+        lineRepository.save(new Line(id, lineUpdateRequest.getName(), lineUpdateRequest.getColor(),
```
`line` 을 조회해왔으니, `line` 에게 정보를 수정하라고 요청하고, 변경이 완료된 엔티티를 저장하는 형태로 구현해보는 건 어떨까요?

aksgudwns replied:
넵 수정하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/domain/Stations.java:
```diff
@@ -0,0 +1,16 @@
+
+import java.util.List;
+
+public class Stations {
```
이 클래스는 사용처가 없는 것 같네요!

aksgudwns replied:
사용하지 않는 클래스 삭제하겠습니다..

seokhongkim reviewed (changes requested):
형준님, 2단계 미션 잘 진행해주셨습니다. 👍 
기능 단위로 인수 테스트를 작성하며 구현을 해주셨네요. 커밋도 기능 단위로 해주셔서 좋았습니다.

코드를 보고 궁금한 점과 고민해보시면 좋을 것 같은 내용을 코멘트로 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

aksgudwns committed:
doc - 작업완료 x표기로 수정

aksgudwns committed:
refactor - 피드백 반영

- 테스트 메서드에 when, then api호출과 검증 명확히 구분하도록 수정
- lineName, stationName 두개 확인 로직 둘다 있을 경우 통과하도록 수정
- create, delete 테스트 생성 api응답값 확인으로 테스트 로직 변경
- getById -> findById 수정 및 메서드 역활 분리
- @Transactional 설정 변경
- 사용하지 않는 클래스 삭제

aksgudwns committed:
refactor - errorMessage enum 분리

aksgudwns commented:
안녕하세요!

말씀주신 내용 반영해보았습니다!
빠르게 피드백 주셔서 감사합니다:)
제가 잘못 이해하거나 보완 필요한 부분 있으면 추가 피드백 부탁드립니다

감사합니다!

seokhongkim commented on README.md:
```diff
@@ -62,22 +62,22 @@ This project is [MIT](https://github.com/next-step/atdd-subway-admin/blob/master
 * 인수 테스트의 결과가 다른 인수테스트에 영향을 끼치지 않도록 인수 테스트를 서로 격리 시키세요.
 * 인수 테스트의 재사용성과 가독성, 그리고 빠른 테스트의도 파악을 위해 인수 테스트를 리팩터링 하세요.
-* [] 지하철노선 생성 
+-[x] 지하철노선 생성 
```
`-` 와 `[` 사이에 공백이 없어서 제대로 적용이 안 된 것으로 보이네요!
> https://lynmp.com/ko/article/title/markdown-list-to811c9dc50n

aksgudwns replied:
수정하겠습니다! 문서까지 링크 같이 주셔서 감사합니다!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -38,26 +35,29 @@ public List<LineResponse> findAllLines() {
+
+    private Station getStationById(Long id) {
+        return stationRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException(ErrorMessage.LINE_NO_FIND_BY_ID.getMessage()));
```
1. 지하철 역 검색에 실패한 것이니 예외 메시지도 역 검색에 실패한 것으로 변경이 필요할 것 같아요.
2. `getStationById`는 지하철 역만 조회하는 메서드이니 `StationService`에서 제공해보면 어떨까요? 비슷한 기능이 필요한 다른 서비스나 컨트롤러가 있다면 `StationService`를 통해 동일한 기능을 재사용할 수 있고, `Station`과 관련된 기능이 모여 있어서 서비스의 응집도가 올라갈 것 같습니다.

aksgudwns replied:
메시지도 알맞게 수정하겠습니다!
해당 메서드는 말씀처럼 StationService에 있는게 적절해 보이는군요.. 수정하겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -78,4 +69,9 @@ public List<StationResponse> getStationResponses() {
                 .collect(Collectors.toList());
     }
+
+    public void updateRequest(LineUpdateRequest lineUpdateRequest) {
```
저는 도메인에서 dto에 의존하는 것(=사용하는 것)을 지양하는데요.
이유는 dto의 변경이 도메인에 영향을 주기 때문입니다.

형준님께서는 어떻게 생각하시는지 궁금하네요!

aksgudwns replied:
객체를 그대로 사용하는 것이 코드가 깔끔할 것 같다고 생각했는데..
말씀듣고보니 도메인 메서드가 dto에 의존하는 것은 바람직해 보이지 않는군요
수정하겠습니다!

seokhongkim commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -85,8 +86,9 @@ void getLinesTest() {
         //then
         assertThat(lineNames.size()).isEqualTo(2);
         assertThat(lineNames.stream()
```
아래와 같이 검증해 볼 수도 있겠네요!
```suggestion        
        Set<String> results = new HashSet<>(lineNames);
        assertThat(results).containsExactly(lineName1, lineName2);
```

aksgudwns replied:
코드가 훨씬 깔끔해졌네요!
StationAcceptanceTest도 같이 수정하겠습니다 :)

seokhongkim commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -137,12 +139,9 @@ void deleteLineTest() {
-        assertThat(retrieveAllLineNames().stream().anyMatch(lineName::equals))
-                .isFalse();
+        assertThat(deleteLine(lineId).statusCode())
+                //then
```
코드 중간에 주석을 추가하는 것보다는 `deleteLine(lineId)` 를 호출하는 부분을 when 구간에 놓고, 변수로 값을 저장한 후 이를 검증하는 부분(`asserThat`)을 then 으로 분리하는 건 어떨까요?

또는 spock 테스트 프레임워크에서는 when, then 을 함께 나타내는 expect 를 사용하기도 합니다.

aksgudwns replied:
호출하는 부분을 분리하는 것이 가독성 측면에서도 더 좋을 것 같습니다!
비슷하게 처리되어 있는 부분들도 말씀대로 수정하겠습니다~

seokhongkim reviewed (approved):
형준님, 2단계 미션 코멘트 잘 반영해주셨습니다. 👍 
2단계 PR도 머지하겠습니다.

코멘트를 조금 더 추가로 남겼는데, 다음 단계 미션 진행하실 때 참고해주셔도 좋겠습니다.
3단계 미션에서 뵙겠습니다. 😀
